File: hackerrank/problem solving/algorithms/2 - implementation/25 - save the prisoner/save_the_prisoner_oo.java
// Source: https://www.hackerrank.com/challenges/save-the-prisoner/problem?isFullScreen=true

import java.util.Scanner;
import java.util.List;
import java.util.Arrays;
import java.util.stream.Stream;
import java.util.stream.Collectors;


public class Solution {

    private static Scanner _scanner;

    public static void main(String[] args) {
        _scanner = new Scanner(System.in);
        final int N_TEST_CASES = _readANumber();
        List<Integer> prisonersChairNumberToWarn = Arrays.asList(new Integer[N_TEST_CASES]);

        for (int i = 0; i < N_TEST_CASES; i++) {
            final List<Integer> TEST_CASE = _readTestCase();
            final SaveThePrisoner OBJ = new SaveThePrisoner(TEST_CASE);
            prisonersChairNumberToWarn.set(i, OBJ.getPrisonerChairNumberToWarn());
        }

        _printResults(prisonersChairNumberToWarn);
    }

        private static int _readANumber() {
            final int NUMBER = _scanner.nextInt();
            _scanner.nextLine(); // remove remaining new line
            return NUMBER;
        }

        private static List<Integer> _readTestCase() {
            final List<Integer> TEST_CASE = Stream.of(_scanner.nextLine().split(" "))
                                            .map(Integer::parseInt)
                                            .collect(Collectors.toList());
            return TEST_CASE;
        }

        private static void _printResults(List<Integer> array) {
            array.forEach(System.out::println);
        }
}

    class SaveThePrisoner {

        private int _prisoners;
        private int _sweets;
        private int _chairNumberToBegin;
        private int _prisonerChairNumberToWarn;

        public SaveThePrisoner(List<Integer> testCase) {
            _prisoners = testCase.get(0);
            _sweets = testCase.get(1);
            _chairNumberToBegin = testCase.get(2);

            _prisonerChairNumberToWarn = _saveThePrisoner();
        }

            private int _saveThePrisoner() {
                int prisonerChairNumberToWarn = _chairNumberToBegin + (_sweets - 1);

                if (prisonerChairNumberToWarn > _prisoners) {
                    prisonerChairNumberToWarn %= _prisoners;

                    if (prisonerChairNumberToWarn == 0)
                        prisonerChairNumberToWarn = _prisoners;
                }

                return prisonerChairNumberToWarn;
            }

        public int getPrisonerChairNumberToWarn() {
            return _prisonerChairNumberToWarn;
        }
    }

File: hackerrank/problem solving/algorithms/2 - implementation/25 - save the prisoner/save_the_prisoner_oo.cpp
// Source: https://www.hackerrank.com/challenges/save-the-prisoner/problem?isFullScreen=true

#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int read_a_number();
vector<int> read_test_case();
void print_results(const vector<int> ARRAY);


class Save_The_Prisioner {

private:
    int _prisoners;
    int _sweets;
    int _chair_number_to_begin;
    int _prisoner_chair_number_to_warn;

    int _save_the_prisoner() {
        int prisoner_chair_number_to_warn {_chair_number_to_begin + (_sweets - 1)};

        if (prisoner_chair_number_to_warn > _prisoners) {
            prisoner_chair_number_to_warn %= _prisoners;

            if (prisoner_chair_number_to_warn == 0)
                prisoner_chair_number_to_warn = _prisoners;
        }

        return prisoner_chair_number_to_warn;
    }

public:
    Save_The_Prisioner(const vector<int> TEST_CASE) {
        _prisoners = {TEST_CASE.front()};
        _sweets = {TEST_CASE.at(1)};
        _chair_number_to_begin = {TEST_CASE.back()};

        _prisoner_chair_number_to_warn = {_save_the_prisoner()};
    }

    int get_prisoner_chair_number_to_warn() const {
        return _prisoner_chair_number_to_warn;
    }
};


int main() {
    const int N_TEST_CASES {read_a_number()};
    vector<int> prisoners_chair_number_to_warn(N_TEST_CASES);

    for (auto &prisoner_chair_number_to_warn : prisoners_chair_number_to_warn) {
        const vector<int> TEST_CASE {read_test_case()};
        const Save_The_Prisioner OBJ(TEST_CASE);
        prisoner_chair_number_to_warn = {OBJ.get_prisoner_chair_number_to_warn()};
    }

    print_results(prisoners_chair_number_to_warn);

    return 0;
}

    int read_a_number() {
        int number;
        cin >> number;
        return number;
    }

    vector<int> read_test_case() {
        const int TEST_CASE_SIZE {3};
        vector<int> test_case(TEST_CASE_SIZE);

        for (auto &number : test_case)
            cin >> number;

        return test_case;
    }

    void print_results(const vector<int> ARRAY) {
        ranges::for_each(ARRAY, [](auto element){cout << element << "\n";});
    }
